msteams-platform/bots/how-to/conversations/send-proactive-messages.md

```diff
@@ -7,6 +7,153 @@ ms.author: anclear
+- **Who did it.** Who took the action that caused the notification to be sent.
+- **What they can do about it.** Make it easy for the users to take actions based on the notification.
+- **How they can opt out.** You need to provide a path for users to opt out of additional notifications.
+
+## Obtain necessary user information
+
+Bots can create new conversations with an individual Microsoft Teams user by obtaining the userâ€™s *unique ID* and *tenant ID.* You can obtain these values using one of the following methods:
+
+> [!WARNING]
+> Please, verify these links; they point to topics in the _old folder.
+
+- By [fetching the team roster](../../../_old/concepts/bots/bots-context.md#fetching-the-team-roster) from a channel your app is installed in.
+- By caching them when a user [interacts with your bot in a channel](../../../_old/concepts/bots/bot-conversations/bots-conv-channel.md).
+- When a users is [@mentioned in a channel conversation](../../../_old/concepts/bots/bot-conversations/bots-conv-channel.md#-mentions) the bot is a part of.
+- By caching them when you [receive the conversationUpdate](../../../_old/concepts/bots/bots-notifications.md#team-member-or-bot-addition) event when your app is installed in a personal scope, or new members are added to a channel or group chat that
+
+### Proactively install your app using Graph
+
+> [!Note]
+> Proactively installing apps using graph is currently in beta.
+
+Occasionally it may be necessary to proactively message users that have not installed or interacted with your bot previously. For example, you want to use the [company communicator](../../../samples/app-templates.md#company-communicator) to send messages to your entire organization. For this scenario you can use the Graph API to proactively install your bot for your users, then cache the necessary values from the `conversationUpdate` event your app will receive upon install.
+
+You can only install apps that are in your organizational app catalogue, or the Teams app store.
+
+See [Install apps for users](https://docs.microsoft.com/graph/teams-proactive-messaging) in the Graph documentation for complete details. There is also a [sample in .NET](https://github.com/microsoftgraph/contoso-airlines-teams-sample/blob/283523d45f5ce416111dfc34b8e49728b5012739/project/Models/GraphService.cs#L176).
+
+
+
+## Example
+
+The following code snippets belong to a bot sample that could be added to a team, but could also work in group chat (with updated `onMembersAdded` implementations).
+If the user `@mention` the bot and sends a message to it, the bot **proactively** sends messages to the user.
+
+Download the complete code example at this location [ProactiveMessages](https://github.com/microsoft/botbuilder-dotnet/tree/master/tests/Teams/ProactiveMessages).
+
+```cs
+protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
```

Comment:
I don't know enough about why they're doing it this way to comment on the sample code.